### ISSUE TITLE
Axe-core shows different issues in Chrome extension and selenium automation

### ISSUE THREAD

**Issue by Sudhi108** [dequelabs/axe-core-maven-html#281]
We are trying to build a automation tool for testing accessibility of our product using axe-core from Selenium. When we did this and ran the test, there are many issues reported from axe-core (12) while on the same page when we use axe chrome browser extension, we don't get MOST of the issues (2). There are 10 issues that are not present in chrome extension.

Also, in each of the violations found, there are something called as nodes. Should these be counted as individual issue? 

for example, there are 10 violations and in each violation there are 5 elements in nodes, then is my total issue count is 10 or 50?

please advice.

**Comment by Sudhi108** [comment]
@Zidious 
Thanks for your replay.

Regarding the first point, I still see difference in the issues reported by the chrome plugin and the axe-core automation via selenium. 
I changed the code to use the withTags method in AxeBuilder and added wcag2a and wcag2aa. Then the number of issues in automation come down. But the issues reported are totally different.

**Comment by Sudhi108** [comment]
@Zidious 

I am building this for our product testing which is not public. However, I built a similar standalone project and put it on github. You can find it here >> https://github.com/Sudhi108/myA11yAutomation.git

Both selenium and the chrome plugin are using 4.4.2 version axe-core.

I have also added the withTags to the builder.

In my code shared here, I am testing www.amazon.com. 
The number of violations i am getting on the axe-core-selenium api is 1 while there are several issues reported on chrome plugin.

In my product testing I have observed the issues reported are totally different.

**Comment by Zidious** [comment]
Hey @Sudhi108 sorry for the delay, I will take a look and report my findings.

**Comment by Sudhi108** [comment]
Hey @Zidious did you get any chance to take a look at this? any findings?

**Comment by Zidious** [comment]
> Hey @Zidious did you get any chance to take a look at this? any findings?

I will be looking into this today

**Comment by Zidious** [comment]
Hey @Sudhi108,

Here are my findings: 


Within your [`AccessbilityBase.java`](https://github.com/Sudhi108/myA11yAutomation/blob/master/src/main/java/testBase/AccessibilityBase.java#L23) class you're only running WCAG2.1A and WCAG2.1AA rules. 

The extension runs WCAG2A, WCAG2AA, WCAG2.1A and WCAG2.1AA rules. To mimic this behaviour you can do the following: 

```java
withTags(Arrays.asList("wcag2a", "wcag2aa", "wcag21a", "wcag21aa"))
```

Using the snippet below, I was able to consistently get 2 violations (1 node in each violation) and quite a few incomplete (needs review in the extension): 

```java
AxeBuilder axeBuilder =
	new AxeBuilder().withTags(Arrays.asList("wcag2a", "wcag2aa", "wcag21a", "wcag21aa"));

WebDriver webDriver = new ChromeDriver(new ChromeOptions().setHeadless(false));

webDriver.get("https://www.amazon.com/");

Results axeResults = axeBuilder.analyze(webDriver);
```

**Comment by Zidious** [comment]
Hey @Sudhi108, 

Are you still having issues? If so, is it possible to provide a reproducible sample that we could use to debug further.

**Comment by Sudhi108** [comment]
hi @Zidious 

I am able to run axe-core with all the tags now. The problem was for older version we had used axe.main.js file which was causing failure. After removing it, we are able to make progress in our automation.

However, I still have extra issues reported in axe-core as compared to chrome plugin. When I looked closely, I found that the issues that needs review are also reported as actual issues by axe-core automation. Is there a way we can identify these in code and eliminate the false positive (meaning the ones that needs review)?

**Comment by Zidious** [comment]
The items within needs review are elements which the axe-core Selenium could not test and need to be reviewed manually. We would need a reproducable test in which we can confirm if this was a false positive or not.

**Comment by Sudhi108** [comment]
@Zidious 

I am using the project I shared earlier and trying to print the violations caught on amazon.com site. On the chrome plugin, i am getting only one issue reported. Using axe-core-selenium automation, I am getting following issues 

Inline text spacing must be adjustable with custom stylesheets serious 1
Elements must have sufficient color contrast serious 51

The color contrast issues reported above are not caught on the plugin. Also the one issue reported by the plugin is different from what is reported by automation. I have also added the screen shots of the plugin with best practices ON and OFF.

You may use my code from https://github.com/Sudhi108/myA11yAutomation.git

![image](https://user-images.githubusercontent.com/115060181/210979524-94e9f225-5a95-482d-ba10-047e394e3b73.png)
![image](https://user-images.githubusercontent.com/115060181/210979794-5a522666-b09d-437b-8943-575db572a940.png)

**Comment by Zidious** [comment]
@Sudhi108,

I'll take another look today.

**Comment by straker** [comment]
@Sudhi108 The devtools extension recently [disabled Needs Review items from showing in the results panel](https://docs.deque.com/devtools-html/4.0.0/en/devtools-4-46-2). I believe this is the reason why the results are not showing up any more (such as color contrast issues). The axe-core apis still have Needs Review turned on, so the results will be different between them in that regard. However, both apis and extensions will report the same issues for all Violations found.

**Comment by Sudhi108** [comment]
@straker Thanks for your inputs. Is there a way I can turnoff the needs review items in axe-core automation or at least identify them? I can do it on chrome plugin so I believe there should be a way we can do it in axe-core also. Please advice.

**Comment by straker** [comment]
I guess it depends on what you're looking at. If you are looking at the entire `Results` object when you save it to a file, there's no way to ignore the `incomplete` (Needs Review) from that as the object shows everything. But based on your code you're only looking at the `violations` for logging and `checkAccessibilityViolations`, so you shouldn't see `incomplete` (Needs Review) in your logging. Are you saying you are seeing ~51 color contrast issues in your console logs?

**Comment by Sudhi108** [comment]
You are right. I am looking at the violations which has 2 entries. One was "Inline text spacing must be adjustable with custom stylesheets" reported on 1 node and second one was "Elements must have sufficient color contrast" reported on 51 nodes. and I have just added print statements to verify them. 

What I am looking for is to somehow identify these 51 needs review type issues from my violations.

**Comment by straker** [comment]
That's strange. We ran the code you provided and do not get the 51 color contrast issues. The code [only outputs violations](https://github.com/Sudhi108/myA11yAutomation/blob/95a66f3259e8830e7351d0374a6d57a61bcc82b3/src/main/java/testBase/AccessibilityBase.java#L61-L63), and only when we switch the loop to look at `getIncomplete()` do we get the color contrast issues (as expected). The results object should already differentiate between `violations` and `incomplete` (needs review) for you.

**Comment by Sudhi108** [comment]
I added these lines in code
```
            for(Rule violation: results.getViolations())
                System.out.println("Violations " + violation.getHelp() + " " + violation.getImpact() + " " + violation.getNodes().size());

            for(Rule incomplete: results.getIncomplete())
                System.out.println("Incomplete " + incomplete.getHelp() + " " + incomplete.getImpact() + " " + incomplete.getNodes().size());
```
and I am getting following results

```
Violations Inline text spacing must be adjustable with custom stylesheets serious 1
Violations Elements must have sufficient color contrast serious 51
Violations Form elements must have labels critical 1
Incomplete Elements must only use allowed ARIA attributes serious 1
Incomplete Elements must have sufficient color contrast serious 44
Incomplete Elements must have sufficient color contrast serious 44
Incomplete Links with the same name must have a similar purpose minor 4
```

**Comment by straker** [comment]
That looks correct, the extension only outputs the violations. So if you're outputting the incomplete results as well you are going to get more issues than the extension currently shows.

**Comment by straker** [comment]
Interesting. I don't get any color contrast issues on amazon. Is your amazon region different? I'm on US amazon and get 0 violations.

**Comment by straker** [comment]
Or perhaps you're signed in to an amazon account when testing in the extension?

**Comment by Sudhi108** [comment]
No. It opens amazon.com site only and not amazon.in. I have verified that. You can see the url in the snapshots attached also

**Comment by straker** [comment]
Odd. We're not able to reproduce the issue and any time we run axe against amazon.com we get 0 color-contrast violations (no matter the version we use). With amazon being such a dynamic site, it could very well be that you are getting content that we are not, and that content has color-contrast violations. Unfortunately, without being to reproduce it on our end I'm not sure there's much we can do to help.

<img width="1680" alt="Screenshot showing Amazon.com with developer tools open. The developer tools show the result of running 'axe.run()' and that there are no color-contrast violations" src="https://user-images.githubusercontent.com/2433219/212935479-61b7e237-543b-4d56-873b-0d154bdc8f88.png">

**Comment by elopezanaya** [comment]
Hi, I hope is ok to jump into this thread,  Im having quite the same issue reported, when running my code it displays 2 errors, but when doing the validation in the browser using dev tools plugin, I got 8 errors. 

The extra 6 errors are coming from an i-frame in my html, so I believe axe-core is not including i-frame in the validation,  even when the iframe validation is enabled, 

this is my line of code , 

` axeResults = await new AxeBuilder({ page }).options({ iframes: true }).withTags(['wcag2a', 'wcag2aa', 'wcag21a', 'wcag21aa']).analyze();
`

this is my HTML -> https://elastorageredmond.blob.core.windows.net/elopez-dev/master/WebChatControl/htmls/chat.html?&id=Microsoft_Omnichannel_LCWidget&data-app-id=920ea4d4-9166-4281-bac7-c89b694baa2a&data-org-id=ed5ece4e-719e-4765-a060-2c30b44ebadd&data-org-url=https://unqed5ece4e719e4765a0602c30b44eb-crm.oc.crmlivetie.com&hostname=elastorageredmond.blob.core.windows.net&data-disable-telemetry=false&data-open-in-window=false


is just an empty HTML with the iframe containing the extra code, 

any help will be welcome,  my apologize in advance in case I need to create another , please let me know.

**Comment by straker** [comment]
@elopezanaya So what I think is going on is that the iframe loading the chat widget is not fully loaded by the time our API code scans the page. Each of the web drivers (playwright, puppeteer, webdriver) will return from the navigation code (`page.goto`, etc.) when the `load` event is fired by the browser. If you run axe right after you navigate then only the page at that state is tested. You'll need to add some amount of timeout before you analyze axe in order to wait for the chat iframe to fully load since the `load` event for that page is fired well before the chat widget is on the page.
